rfcs/text/0000-matrix-service.md

```diff
@@ -0,0 +1,418 @@
+
+The Matrix Service also does not have a migration strategy for its index and requires a full re-index to respond to metadata schema changes.
+
+#### Delete/Redact
+
+Deletion and redaction are handled the same way and are straightforward processes for the Matrix Service:
+
+1. Remove all data from Redshift that is associated with bundle FQIDs from a given project UUID
+2. Delete project matrices from Data Brower's S3 bucket
+3. Clear the [Matrix cache](#matrix-cache)
+
+This process is performed manually by a Matrix Service developer to ensure correct execution; it is anticipated that this process will occur infrequently. On project redaction, DataOps will open a Matrix Service ticket to request and track the process. The team adheres to a 1-week period to purge all associated data with redacted/deleted projects.
+
+### Availability
+
+#### Request ID
+
+The `request_id` returned by `POST /matrix` is valid for 30 days following the successful completion of the request. Specifically, this refers to the S3 download URL to retrieve the generated matrix will be valid for 30 days. Following this time, matrices are deleted from S3 and `GET /matrix/request_id` will notify clients that the request has expired.
+
+#### Matrix cache
+
+Matrices are cached in S3 according to a hash of a [request's input parameters](#post-matrix). Specifically, the hash encodes the set of `cellkey`s selected by the user's `filter` object, the list of metadata `fields` exported, the output `format`, and `feature`. During a request, matching hashes short-circuit the matrix generation process for which a copy of the existing matrix is created for this request. This enables matrices to return within minutes, if not seconds, where the bottleneck of the request becomes the copy operation of the matrix.
+
+Matrices are cached for 30 days following the most recent cache hit. The cache is cleared when data in the index is updated (e.g. project reingestion, bundle updates).
+
+#### Project matrices
+
+Project matrices served via the Data Browser are regenerated and updated when projects are reingested or relevant bundle updates are applied to the index.
+
+### Reproducibility
+
+The Matrix Service will maintain a single version of a bundle for which will be the latest version that belongs to a [version-complete](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0015-project-completion-stage.md#definitions) project. The following reproducibility implications apply:
+
+1. Matrices for historical versions of bundles/projects can not be generated.
+2. The Matrix Service is non-deterministic over time. Identical parameters supplied to the Matrix Service at two different times does not guarantee identical outputs. 
+3. The current Matrix Service does not support an implementation for Citations or Snapshots.
+
```

Comment:
Agreed that the Matrix Service is non-deterministic and that reproducibility would be out of scope for Matrix Service. However, just from the perspective of a user that potentially will generate many matrices with multiple filters in a short amount of time, it is easy to get confused and tedious to organize as to which matrix corresponds with which post request. I remember this was brought up before, but would it possible to add logs (not to imply reproducibility but rather for historical logging purposes)?

Reply:
Yes, I've run into a similar pattern and there isn't a great way to manage it. Could you expand on what you had in mind for logs?

Would returning information about your `POST` request in the `GET` request be helpful? I'm imagining a `verbose` or `info` parameter on the `GET` endpoint could include the initial `filter` object in the response.

Reply:
I think returning info about the POST in the GET request would definitely be helpful. Another possibility is to actually create a log file along with the downloaded matrix files. The con for that situation would be that that would be another file added to the download, and for mtx, that potentially is already a lot of files.

Reply:
Hm yea, we have a similar issue with `loom` in which we're unable to provide additional files (e.g. summary stats, README) as users are expecting to consume the file directly without having to unzip it.